File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/dashboard/presenter/MyDashboardPresenterImpl.java
package org.wikiedufoundation.wikiedudashboard.dashboard.presenter;

import android.util.Log;

import org.wikiedufoundation.wikiedudashboard.dashboard.data.MyDashboardResponse;
import org.wikiedufoundation.wikiedudashboard.dashboard.provider.MyDashboardProvider;
import org.wikiedufoundation.wikiedudashboard.dashboard.view.MyDashboardView;
import org.wikiedufoundation.wikiedudashboard.helper.PresenterCallback;

public class MyDashboardPresenterImpl implements MyDashboardPresenter {
    private MyDashboardView myDashboardView;
    private MyDashboardProvider myDashboardProvider;

    public MyDashboardPresenterImpl(MyDashboardView myDashboardView, MyDashboardProvider myDashboardProvider) {
        this.myDashboardView = myDashboardView;
        this.myDashboardProvider = myDashboardProvider;
    }

    @Override
    public void requestDashboard(String cookies) {
        myDashboardView.showProgressBar(true);
        myDashboardProvider.requestCourseList(cookies, new PresenterCallback() {
            @Override
            public void onSuccess(Object o) {
                myDashboardView.showProgressBar(false);
                MyDashboardResponse myDashboardResponse = (MyDashboardResponse)o;
                Log.d("Presenter: ", myDashboardResponse.toString());
                myDashboardView.setData(myDashboardResponse);
            }

            @Override
            public void onFailure() {
                myDashboardView.showProgressBar(false);
                myDashboardView.showMessage("Unable to connect to server.");

            }
        });
    }
}

File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/ui/dashboard/MyDashboardPresenterImpl.kt
package org.wikiedufoundation.wikiedudashboard.ui.dashboard

import org.wikiedufoundation.wikiedudashboard.ui.dashboard.data.MyDashboardResponse
import org.wikiedufoundation.wikiedudashboard.ui.dashboard.data.UserData
import org.wikiedufoundation.wikiedudashboard.util.PresenterCallback
import timber.log.Timber

/**
 * Class to implement [requestDashboard]
 * @constructor primary constructor
 *
 * @property myDashboardView view component for course list
 * @property myDashboardProvider retrofit HTTP request call for course list
 * ***/
class MyDashboardPresenterImpl(
    private val myDashboardView: MyDashboardContract.View,
    private val myDashboardProvider: MyDashboardContract.Provider
) : MyDashboardContract.Presenter {

    override fun requestDashboard(cookies: String) {
        myDashboardView.showProgressBar(true)
        myDashboardProvider.requestCourseList(cookies, object : PresenterCallback<MyDashboardResponse> {
            override fun onSuccess(response: MyDashboardResponse) {
                myDashboardView.showProgressBar(false)
                Timber.d(response.toString())
                myDashboardView.setData(response)
            }

            override fun onFailure() {
                myDashboardView.showProgressBar(false)
                myDashboardView.showMessage("Unable to connect to server.")

                val unknownDashboardResponse =
                        MyDashboardResponse(
                                UserData(""), ArrayList()
                        )
                myDashboardView.setData(unknownDashboardResponse)
            }
        })
    }
}
